### ISSUE TITLE
[Plugin] Service API documentation

### ISSUE THREAD

**Issue by stefanalund** [spotify/backstage#627]
## Summary

The following illustration shows how a future potential solution for API discoverability and API documentation could look in Backstage. The initial focus is Service APIs.

![Screenshot 2020-04-23 at 11 46 26](https://user-images.githubusercontent.com/190856/80085145-2144e180-8558-11ea-8e0e-1e6fb6a51348.png)

* In the left-hand column, API development, engineers build their API (gRPC API with Protocol Buffers or REST API with Open API) together with the code.
* In the right-hand column, Discovery, you have three ways to discover: 1) Through the Service 2) Using a catalogue to **browse** APIs by use case 3) Through Search. This plugin should expose 1 and 2. 
* In the middle column, API user experience, you are in Backstage and you select your API version. The API is shown in a Swagger-like manner and you are able to interact with the API endpoints.

For inspiration, inspecting the API of a backend service is shown in this short video: https://youtu.be/vcDL9tOv7Eo?t=80

## Project websites (if applicable)

https://swagger.io/docs/specification/about/
https://developers.google.com/protocol-buffers
https://github.com/thiagobustamante/typescript-rest-swagger

## Context

API documentation should be a first class citizen in the Backstage system model: https://github.com/spotify/backstage/issues/390

**Comment by aexvir** [comment]
Hi @stefanalund , we already have this feature implemented for OpenAPI, although I'm a bit confused how this could be implemented into Backstage.

Right now we are repurposing our backend to be feeding a Backstage frontend, and we'll have this feature straight away, but it would require our service discovery, and basically our backend to be running. We could definitely separate it into its own backend, and easily provide an API overview page, but then I'm not sure how we could also integrate it to our service page.

As far as I recall, you mentioned that in that case the information passing from one backend to other would be taken care in a separate layer, but it's still a bit blurry, could you explain a bit further?

I can see this working if we'd have a OpenAPI plugin for example, and then a dependency in our Services plugin, and only in that case we'd display the specific OpenAPI definitions of a service in their detail page, is that how it's supposed to work? Or is there a better approach?

**Comment by stefanalund** [comment]
@aexvir it is fussy to me as well :) When we get further on the Backstage backend architecture, I think it will be easier to see the path forward. Ping @freben @mbruggmann

**Comment by freben** [comment]
This is super interesting! All of what you say sounds familiar and are problems that fit well under the Backstage umbrella. I won't be able to write up a response right now, but I'll get back to it as soon as possible!

**Comment by stefanalund** [comment]
> As the owner of an API, it would be nice to integrate linting tools like Spectral to show me the quality level of my APIs.

I really like this. Could see it as something that you would run in the background and show any warnings/errors in the Backstage plugin. This is a common pattern for Backstage at Spotify; we use visualisations in Backstage as a way to inform and nudge engineers towards following best practices.

**Comment by RoyJacobs** [comment]
It's one of the things I implemented in our API Registry, but it led to all sorts of unwanted complexity (embedding a Node application like Spectral in a Kotlin app, etc). A Backstage plugin would be much better suited for this. Frontend editing components like CodeMirror even have a very straightforward way of [visualizing the linting results](https://codemirror.net/demo/lint.html).

**Comment by mbruggmann** [comment]
Interesting, I think there are a few aspects to unpack here:
* Some type of registry or inventory of which APIs exist, their type (OpenAPI, GraphQL, gRPC), the component/system/repo they belong to, and where to query their schema. This could possibly live with Backstage?
* Indexing the actual API schemas, either from source code, through reflection at runtime, or through a manual process. Could be specific to the type of APIs and to whatever process is used by the company. This is where I see tools such as Katalog coming in.
* UX for Searching, Browsing and Visualizing APIs. Probably a good fit or Backstage

**Comment by RoyJacobs** [comment]
Yes, that seems about right. Katalog could then be reduced in scope quite a bit to 'just' host the API specs and providing a flexible query API tor retrieve them.

One thing that would have to be carefully considered is how the link is made with the metadata that is stored in Backstage: Is the service name in Backstage a unique key under which Katalog can then store API specification files? How does it work when a service is deleted, etc?

**Comment by tkrop** [comment]
In Zalando, we implemented this feature in a standalone API Portal supported by the API Repository providing a historic perspective on all deployed APIs. The API repository is supporting an easy search language that directly provides results from an ElasticSearch engine.

Content is pushed to the API Repository via a job running every 5 minutes to collect all current deployments and resolve their APIs from two potential source:

1. The Docker image via a service called API Scanner which is working as a big cache, or 
2. as fallback by querying the .well-known/schema-discovery endpoint implemented by the application.

We prefer the first source, as it provides a more precise picture in case of multiple application stacks, however, we have gateways publishing their dynamic target APIs through the legacy schema-discovery. We actually would prefer them to push this information directly to the API repository, but we have not provided the API for this yet.

We although have some quality metrics for APIs based on our API review database as well as our API linter [Zally](zalando/zally) that we show in the API Portal.

In the API Portal we are today a bit stuck from technology perspective. Adapting the Swagger library was challenging 3 years ago and we introduced some dependencies that are not maintained any longer and hard to substitute. However, the backend is very simple and robust. May be we can contribute with our experience and provide some open source components here.

**Comment by stefanalund** [comment]
@tkrop thanks for the write-up!

Is the API Portal (and perhaps API Scanner) open source or something you would consider releasing as part of Backstage? Sounds like it could be a good match. 

I imagine that API scanning and ingestion step may be very dependent on a company's underlying infrastructure, thus making it hard to provide an end-2-end solution. Backstage could provide a standardised UI (through a plugin) and a backend with a clear ingestion API.

**Comment by tkrop** [comment]
@stefanalund I think we are willing to contribute, if the community is interest enough.

Interestingly, the API scanner is the part to open source easiest while, we think, already having a very high value for the community. It has nearly no company specific hooks, and the existing ones can be removed or disabled easily. It has only very small technical debts and provides a good potential for extension by the community, e.g. by supporting other API specification and documentation formats. A simple backstage plugin could already integrate against the API scanner API to visualize APIs based on Docker images names.

Next easiest would be to opensource the API repository consisting of a Postgres and ElasticSearch data storage for providing the historical API information. With relatively small effort, we could remove the legacy API and integrate the search feature from the API Portal to make it usable for the community. While the data model is sufficiently generic it is a still a bit incomplete, since we have not developed the whole envisioned model. Also the API has some rough edges and supports only a small number of use cases yet. Anyhow, nothing that cannot be fixed by the community.

To make this work, users need to implement a discovery job/service, that populates the API Repository using the data from the API Scanner. This is the company specific part of our architecture that also offers the freedom of adding APIs not detectable via the API Scanner. Finally, a backstage plugin needs to integrate against the API Repository search API to search and present the historic API data.

**Comment by RoyJacobs** [comment]
@tkrop It would be great if the scanner could be optional. At bol.com we are aiming for build-time linting and registration of APIs (i.e. before the service goes live, to prevent any accidents). For JVM services we could then provide a Maven plugin that performs the scan in the local project before uploading to your API repository.

Second, perhaps it would be nice if the Elasticsearch dependency is optional and it could be made to work just with the Postgres full-text search functionality. ES is quite heavyweight and it would increase some operational burden. This can also be a community effort of course, though :)

**Comment by stefanalund** [comment]
@tkrop I'm know for a fact that there are several companies that would be interested in this. Also, speaking on behalf Spotify, we don't have a good story for API documentation and would gladly adopt something like what you are proposing 👍

**Comment by tkrop** [comment]
@RoyJacobs yes, build time linting is supported by [zally](https://github.com/zalando/zally).

**Comment by RoyJacobs** [comment]
@tkrop Ah, great! We'll finally start a pilot project with Backstage soon and that will definitely be on my list, then :)

**Comment by stefanalund** [comment]
@freben @Rugvip @mbruggmann now that the [service description file format](https://github.com/spotify/backstage/blob/master/docs/architecture-decisions/adr002-default-catalog-file-format.md#format) is defined, could you provide an example (in this issue) for how you would link up a OpenAPI spec yaml file with the component?

For everyone else in this thread: would you be willing to take on building this plugin? It should be visible on the service (component) page and could use the open source Swagger UI (as demonstrated by @aexvir above):

![Screenshot 2020-06-26 at 14 07 40](https://user-images.githubusercontent.com/190856/85855429-8ca86b00-b7b6-11ea-8b57-ecd972631c6c.png)

We are starting work to create an [end-2-end demo](https://github.com/spotify/backstage/milestone/14) of Backstage with a few key plugins, and I think this one should be part of it.

**Comment by Rugvip** [comment]
@freben @mbruggmann, what do you think about starting out simple with `backstage.io/api-url` + `backstage.io/api-type` annotations or something similar? I'm thinking local API descriptions might make sense for individual services anyway, even with a separate `Api` kind that are attached to systems

**Comment by mbruggmann** [comment]
@Rugvip Yeah, starting with just having the type and url sounds good. Though I would probably have gone with a separate kind, and then link them to components through the name. Eg this:
```yaml
---
apiVersion: backstage.io/v1alpha1
kind: Component
metadata:
  name: frobs-awesome
spec:
  type: service
  implements-apis: # <-- not sure how to call this, nor if it should be an annotation/label instead
    - frobs-awesome-api
---
apiVersion: backstage.io/v1alpha1
kind: API
metadata:
  name: frobs-awesome-api
spec:
  type: openapi
  url: https://...
```
Mostly because I consider the Component to API relationship many-to-many.

**Comment by Rugvip** [comment]
@mbruggmann Yep, agree that we should go with the from the beginning if it doesn't snowball into a bunch of other work and problem solving.

**Comment by freben** [comment]
The only remaining question then is, how does this Api kind entity enter the catalog? It would probably not be convenient to have to specify it in its entirety, if the ingestor can find files that are next to the entity file (or pointed to, by relative path, by the entity file).

**Comment by tkrop** [comment]
@Rugvip @mbruggmann what is it you expect to be provided in the `api-url`?

**Comment by larsnaesbye** [comment]
This seems like it would Backstage work as external developer portal too. I fully applaud this and I think a lot of people would like to get on board Backstage if they could see one solution working as both. We at Kortforsyningen would certainly do it.

**Comment by RoyJacobs** [comment]
Regarding the location of the API, for a lot of our current applications the OpenAPI spec file doesn't exist in the Git repository as-is. Instead, it is generated during build time by introspecting annotations on the Java code (using tools like [this](https://github.com/openapi-tools/swagger-maven-plugin)).

So by itself the API spec file is not scrapable(*) from the Git repo but if the `api-url` could point to any arbitrary URL it could point either to the service itself exposing the API spec file via a well-known url, or it could point to some other platform service we'd have to build ourselves that knows how to extract the API spec file from a build artifact (a JAR in our case).

(*) Is that even a word? :)

**Comment by Rugvip** [comment]
@tkrop We'd likely start by doing OpenAPI schemas e2e, so the url is probably just the location of a schema file. How to support private repositories I think we can expect to come out of work like https://github.com/spotify/backstage/issues/1565

@RoyJacobs yep, could use a URL in the API entity, but I think there could also be an opportunity to have a custom processor in the catalog that is able to import all APIs known to the service.

**Comment by tkrop** [comment]
The `api-url` pointing to a well-known endpoint could be one method. However, at Zalando we discarded this some time ago and instead copy API specs into a well-known directory in the container - which also works nicely with generating API specs. The only thing you can't do is merging APIs dynamically on a server and publish it. We can also point to the `container-api-scanner` that acts as a cache for APIs found in Docker images.

I must admit, I would love to see a solution that allows to calculate or lookup `api-url`s to make these config files less maintenance extensive preventing ot writing code that writes config files.

**Comment by Fox32** [comment]
Hello everyone, we (@SDA-SE) are interested in making progress for the Service API documentation plugin. There isn't anybody working on it, right? We can provide a little MVP for the plugin. But to make sure we are on the same level, here a short recap what we think we should build:

* As [ADR-005](https://github.com/spotify/backstage/blob/master/docs/architecture-decisions/adr005-catalog-core-entities.md) mentions APIs to be core entities, we would expect that they are part of the catalog model and backend. Otherwise it could be a separate plugin that is not part of the catalog. Any suggestion for names? One idea would be *api-docs* or *api-catalog*, however both are a bit close to existing plugins.
* The storage is done in the catalog backend. It has a name, a type and stores a copy of the spec so it's always reachable in Backstage. For example:
```
apiVersion: backstage.io/v1alpha1
kind: API
metadata:
  name: frobs-awesome-api
spec:
  type: openapi/asyncapi/grpc
  definition: |
    openapi 3.0.0
    ... string with the full spec
```
* There is no perfect solution for discovery/ingestion that fits every need. We would like to have the same location and processor system as the service catalog has. This allows everyone to have a flexible ingestion from an URL or file system, during build time or run time... However, for the start we would expect that the location references a API entity and not a raw API spec (like an OpenAPI file).
* We would start with OpenAPI and are interested to add support for AsyncAPI, too.
* Beside the standalone API entity, linking to services in the catalog is important. We would expand the Component entity by an opertional `implementsApis` array that references the APIs by name. For example:
```
apiVersion: backstage.io/v1alpha1
kind: Component
metadata:
  name: frobs-awesome
spec:
  type: service
  implementsApis:
  - frobs-awesome-api
  - other-api
```
* For the separate frontend plugin: First, a list of all APIs, a detail page where one can view an API. The API viewer depends on the type. We start with OpenAPI using swagger-ui and a simple fallback to text output for other formats. Second, we extend the service catalog plugin to show the APIs that a component implements. In the first step we we will not make it very extensible, but always include it directly (like the sentry plugin).

Even though we think these things are interesting, we will skip them for now:
* Other spec formats, like gRPC, as we are missing expertise here
* History or versioning of APIs (we only have "latest")
* Searching for APIs or even better indexing the specs itself.

We would like to have some feedback, but would also directly start with an implementation. With a PR we have some ground for discussions.

**Comment by dtuite** [comment]
Sounds like a good idea to descope search for the moment. There will need to be a backstage-wide effort to address that.

**Comment by RoyJacobs** [comment]
Great, @Fox32! Once this is in place I think we would be open to contributing the history/versioning part. Probably including an API to query based on NPM or Ivy-style semver strings (e.g. things like `^1.0.4`).

**Comment by Fox32** [comment]
Now that #1737 is (nearly) merged, here is what we identified as open topics:
* How dependencies from component to API are modeled (own entity or as part of the component?)
* Modeling consumes relationships
* From the [full model, Systems can expose](https://github.com/spotify/backstage/blob/master/docs/features/software-catalog/system-model.md#concepts) APIs, is this an implicit or an explicit relationship? (On the other hand Domains and Systems are not part of Backstage yet, but we are interested to extend this!)
* How to model versions of APIs/components?
* More API types like gRPC, GraphQL, ...
* Better/Other ingestion stories. Requiring the API definition to be nested into a YAML entity isn't optimal because one often has the API definition itself in source control. So loading them directly would be desirable.

We won't start with everything right now, but will pick something that we think would be a good fit to continue.

**Comment by freben** [comment]
Regarding ingestion, I wonder if there's a story to be had about referencing into other systems, that are specialised for ingestion and versioning and analysis and field searching etc of APIs. Then the question becomes, does a Component entity reference into that somehow, or does an API entity do that? Does the API entity somehow mirror that actual contract data as well or not?

**Comment by nikek** [comment]
As #1737 is merged, I will close this issue. Please feel free to open new issues for the additional improvements.

**Comment by brethubbard** [comment]
I know this is an old issue but these were identitfied as open topics.

> Now that #1737 is (nearly) merged, here is what we identified as open topics:
> 
> * How dependencies from component to API are modeled (own entity or as part of the component?)
> * Modeling consumes relationships
> * From the [full model, Systems can expose](https://github.com/spotify/backstage/blob/master/docs/features/software-catalog/system-model.md#concepts) APIs, is this an implicit or an explicit relationship? (On the other hand Domains and Systems are not part of Backstage yet, but we are interested to extend this!)
> * How to model versions of APIs/components?
> * More API types like gRPC, GraphQL, ...
> * Better/Other ingestion stories. Requiring the API definition to be nested into a YAML entity isn't optimal because one often has the API definition itself in source control. So loading them directly would be desirable.
> * Improve styling of embedded react components (Swagger UI, AsyncAPI), especially for dark mode.
> * Small UX improvement: Button to access the raw API definition in the UI, e.g. to pass it to code generators.
> 
> We won't start with everything right now, but will pick something that we think would be a good fit to continue.

Were these tracked anywhere else? Specifically I am interested in "How to model versions of APIs/components?" I've tried my best to sift through other issues but I can't find anything else besides this thread.

**Comment by Rugvip** [comment]
@brethubbard we actually just added some docs with thoughts on that topic! :grin:

https://backstage.io/docs/features/software-catalog/extending-the-model#referencing-different-environments-with-the-model

TL;DR: Different major versions of an API are different APIs, other than that we only supply tools for modelling a single (latest) version right now.

**Comment by brethubbard** [comment]
Oh perfect, that answers my questions nicely! Thank you!

**Comment by dweber019** [comment]
> [sagacity](/sagacity)

There is now a spectral plugin https://github.com/dweber019/backstage-plugin-api-docs-spectral-linter 😉
